### PR TITLE
Force synchronous upload and reuse of possibly modified spawn outputs

### LINKED ISSUE #22501: Remote cache warning when using `--experimental_remote_cache_async`
### Description of the bug:

When running tests with the remote cache enabled and passing `--experimental_remote_cache_async` I often see this warning:

```
WARNING: Remote Cache: .../1/test.err (No such file or directory)
WARNING: Remote Cache: .../2/test.err (No such file or directory)
WARNING: Remote Cache: .../3/test.err (No such file or directory)
WARNING: Remote Cache: .../4/test.err (No such file or directory)
WARNING: Remote Cache: .../5/test.err (No such file or directory)
```

Where the granularity is per test target. I don't think it shows for every test target but I can see dozens in a single build testing ~600 targets.

### Which category does this issue belong to?

_No response_

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

_No response_

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

7.1.1

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### LINKED ISSUE #23288: Race conditions with local action deduplication
While testing path mapping on Bazel itself with 7.3.0, I just encountered the following exceptions:
```
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData0{actionLookupKey=ConfiguredTargetKey{label=//src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller:v2, config=BuildConfigurationKey[153e20f2e3739e0fdcb3b5e06c3ec8614ec4e43b5ff5f93c78bf2a56108ee671]}, actionIndex=0}' (requested by nodes 'ArtifactNestedSetKey[4]@42667357', 'ArtifactNestedSetKey[3]@612534160', 'ArtifactNestedSetKey[8]@776188326')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalStateException: Missing original path for mapped path bazel-out/darwin_arm64-fastbuild/bin/src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller/libcontroller-hjar.jar in bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller/libv2-hjar.jdeps
jdeps: dependency {
  path: "bazel-out/darwin_arm64-fastbuild/bin/src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller/libcontroller-hjar.jar"
  kind: EXPLICIT
}
rule_label: "//src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller:v2"
success: true

path map: {bazel-out/cfg/bin/external/rules_java~/toolchains/platformclasspath.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_java~/toolchains/platformclasspath.jar, bazel-out/cfg/bin/src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller/libcontroller-hjar.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/src/main/java/com/google/devtools/build/lib/sandbox/cgroups/controller/libcontroller-hjar.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/errorprone/error_prone_annotations/2.24.1/header_error_prone_annotations-2.24.1.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/errorprone/error_prone_annotations/2.24.1/header_error_prone_annotations-2.24.1.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/errorprone/error_prone_type_annotations/2.23.0/header_error_prone_type_annotations-2.23.0.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/errorprone/error_prone_type_annotations/2.23.0/header_error_prone_type_annotations-2.23.0.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/github/stephenc/jcip/jcip-annotations/1.0-1/header_jcip-annotations-1.0-1.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/github/stephenc/jcip/jcip-annotations/1.0-1/header_jcip-annotations-1.0-1.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/code/findbugs/jsr305/3.0.2/header_jsr305-3.0.2.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/code/findbugs/jsr305/3.0.2/header_jsr305-3.0.2.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/guava/guava/33.0.0-jre/header_guava-33.0.0-jre.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/guava/guava/33.0.0-jre/header_guava-33.0.0-jre.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/guava/failureaccess/1.0.2/header_failureaccess-1.0.2.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/guava/failureaccess/1.0.2/header_failureaccess-1.0.2.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/header_listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/header_listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/com/google/j2objc/j2objc-annotations/2.8/header_j2objc-annotations-2.8.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/com/google/j2objc/j2objc-annotations/2.8/header_j2objc-annotations-2.8.jar, bazel-out/cfg/bin/external/rules_jvm_external~~maven~maven/org/checkerframework/checker-qual/3.42.0/header_checker-qual-3.42.0.jar=bazel-out/darwin_arm64-opt-exec-ST-fad1763555eb/bin/external/rules_jvm_external~~maven~maven/org/checkerframework/checker-qual/3.42.0/header_checker-qual-3.42.0.jar}
	at com.google.devtools.build.lib.rules.java.JavaCompileAction.createFullOutputDeps(JavaCompileAction.java:796)
	at com.google.devtools.build.lib.rules.java.JavaHeaderCompileAction.afterExecute(JavaHeaderCompileAction.java:147)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:262)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1159)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1076)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:573)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:862)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:334)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:172)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	... 7 more
```

and

```
java.io.FileNotFoundException: /private/var/tmp/_bazel_fmeum/507738cfc7e6cde00e4a0230e9aa0722/execroot/_main/bazel-out/_tmp/actions/remote/175.tmp (No such file or directory)
	at com.google.devtools.build.lib.unix.NativePosixFiles.lstat(Native Method)
	at com.google.devtools.build.lib.unix.UnixFileSystem.statInternal(UnixFileSystem.java:212)
	at com.google.devtools.build.lib.unix.UnixFileSystem.stat(UnixFileSystem.java:201)
	at com.google.devtools.build.lib.vfs.Path.stat(Path.java:290)
	at com.google.devtools.build.lib.vfs.FileSystemUtils.moveFile(FileSystemUtils.java:456)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.moveOutputsToFinalLocation(RemoteExecutionService.java:878)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.waitForAndReuseOutputs(RemoteExecutionService.java:1473)
	at com.google.devtools.build.lib.remote.RemoteSpawnCache.lookup(RemoteSpawnCache.java:169)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:151)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:118)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.rules.java.JavaCompileAction.execute(JavaCompileAction.java:425)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1155)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1068)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:166)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:95)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:559)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:902)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:349)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:190)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:467)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

_Originally posted by @fmeum in https://github.com/bazelbuild/bazel/discussions/22658#discussioncomment-10330250_

### PR DESCRIPTION
**Pull request by fmeum**
When an action may modify a spawn's outputs after execution, the upload of outputs to the cache and reuse for deduplicated actions need to happen synchronously directly after spawn execution to avoid a race.

This commit implements this for cache uploads by marking all actions with this property and simply disabling async upload for all spawns executed by such actions.

For output reuse, all executions deduplicated against the first one register atomically upon deduplication and cause the cache upload to wait for all of them to complete reuse. 

Fixes #22501
Fixes #23288 
Work towards #21578 
Closes #23307 (no longer needed)

### CODE REVIEW

**Comment by fmeum** [comment]
FYI @aranguyen: We don't need to modify anything about how paths are rewritten in Java actions to benefit from execution deduplication with this change. We should still look into getting rid of the in-place modification for direct classpath header compilation to avoid the synchronous fallback introduced by this change.

**Comment by fmeum** [comment]
@tjgq I have no idea how to test this short of running a large build in CI. I tested it manually on Bazel itself, with promising results: I no longer get any failures and build times improve substantially.

Representative build results for `bazel build //src/main/java/...` with an initially empty disk cache on an M3 MacBook:
```
# Baseline
INFO: Elapsed time: 102.879s, Critical Path: 24.86s
INFO: 5128 processes: 145 internal, 3076 darwin-sandbox, 1907 worker.

# --experimental_worker_multiplex_sandboxing
INFO: Elapsed time: 115.551s, Critical Path: 29.83s
INFO: 5128 processes: 145 internal, 3076 darwin-sandbox, 1907 worker.

# --experimental_worker_multiplex_sandboxing --experimental_output_paths=strip, with deduplication patched out
INFO: Elapsed time: 95.024s, Critical Path: 27.01s
INFO: 5128 processes: 1438 disk cache hit, 145 internal, 2375 darwin-sandbox, 1170 worker.

# --experimental_worker_multiplex_sandboxing --experimental_output_paths=strip
INFO: Elapsed time: 80.242s, Critical Path: 22.04s
INFO: 5128 processes: 1367 disk cache hit, 145 internal, 2171 darwin-sandbox, 459 deduplicated, 986 worker.
```

I expect the sandboxing overhead to be lower on Linux, but even with the overhead, path mapping is substantially faster.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnCache.java [hunk 1]
```diff
@@ -109,89 +109,112 @@ public CacheHandle lookup(Spawn spawn, SpawnExecutionContext context)
...
-              .setFetchTimeInMs((int) fetchTime.elapsed().toMillis())
-              .setTotalTimeInMs((int) totalTime.elapsed().toMillis())
-              .setNetworkTimeInMs((int) action.getNetworkTime().getDuration().toMillis());
-          SpawnResult spawnResult =
-              createSpawnResult(
-                  digestUtil,
+        thisExecution = LocalExecution.createIfDeduplicatable(action);
+        if (shouldUploadLocalResults && thisExecution != null) {
+          LocalExecution previousOrThisExecution =
+              inFlightExecutions.merge(
```
Comment (fmeum) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnCache.java:116]: I *think* that we don't need to this atomically, we could CAS into the map and then register as a second step. But I'm also not 100% sure and I guess this lambda is fast enough that it doesn't hurt.

**Review by tjgq — CHANGES_REQUESTED** [review]
This is great! I think we should be in good shape to flip `--experimental_remote_cache_async` once it's in.

**Review by tjgq — CHANGES_REQUESTED** [review]
I'm not going to insist too much because I know this might be a lot of effort, but: is there some way we can write a test to verify that a fast, async-unsafe spawn waits for a slow, concurrent one to reuse its outputs before proceeding?

**Comment by fmeum** [comment]
> I'm not going to insist too much because I know this might be a lot of effort, but: is there some way we can write a test to verify that a fast, async-unsafe spawn waits for a slow, concurrent one to reuse its outputs before proceeding?

I added a Java test that relies on mocking to make `waitForAndReuseOutputs` controllably slow. I don't think it's guaranteed to catch the issue if it comes back, but I verified that removing the call to `awaitAllOutputReuse` makes it fail. 

The problem is that we can't just make the other spawn slow (it doesn't run anyway), we need to make its output reuse slow. A more realistic test would be possible if we could somehow delay file system operations. Is that something that the test filesystem supports (say, stall updates to `/exec/root/bazel-bin/k8-opt/bin/output` on a synchronization barrier)?

**Comment by tjgq** [comment]
> > I'm not going to insist too much because I know this might be a lot of effort, but: is there some way we can write a test to verify that a fast, async-unsafe spawn waits for a slow, concurrent one to reuse its outputs before proceeding?
> 
> I added a Java test that relies on mocking to make `waitForAndReuseOutputs` controllably slow. I don't think it's guaranteed to catch the issue if it comes back, but I verified that removing the call to `awaitAllOutputReuse` makes it fail.
> 
> The problem is that we can't just make the other spawn slow (it doesn't run anyway), we need to make its output reuse slow. A more realistic test would be possible if we could somehow delay file system operations. Is that something that the test filesystem supports (say, stall updates to `/exec/root/bazel-bin/k8-opt/bin/output` on a synchronization barrier)?

It might be possible with a custom FileSystem implementation; see https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/remote/ActionInputPrefetcherTestBase.java;l=134;drc=3fe227ff418ef81eb5f221625bfd90f52ac70150 for inspiration.

**Comment by fmeum** [comment]
Thanks for the pointer, that looks pretty interesting. Since the main code path here goes through `FileSystemUtils.copyFile` and not a fixed method on `Path`, I think that the current approach may even be more robust though (I am looking into changing the implementation to use NIO).

I added an assert that should make it less likely that the test fails open.

File: src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java [hunk 2]
```diff
@@ -755,6 +765,112 @@ public void pathMappedActionIsDeduplicated() throws Exception {
...
+                throw new IllegalStateException(e);
+              }
+            });
+    completeFirstSpawn.start();
+    // Make it more likely to detect races by waiting for this thread to make as much progress as
+    // possible before letting the second spawn continue.
+    while (completeFirstSpawn.getState().compareTo(Thread.State.WAITING) < 0) {
+      Thread.sleep(10);
+    }
+    assertThat(completeFirstSpawn.getState()).isEqualTo(Thread.State.WAITING);
```
Comment (tjgq) [inline src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java:855]: Can we instead use a latch to explicitly synchronize with the call to `uploadOutputs` (and avoid unnecessarily slowing down the test)?

File: src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java [hunk 3]
```diff
@@ -755,6 +765,112 @@ public void pathMappedActionIsDeduplicated() throws Exception {
...
+
+    RemoteExecutionService remoteExecutionService = cache.getRemoteExecutionService();
+    CountDownLatch enteredWaitForAndReuseOutputs = new CountDownLatch(1);
+    CountDownLatch completeWaitForAndReuseOutputs = new CountDownLatch(1);
+    Mockito.doAnswer(
+            (Answer<SpawnResult>)
+                invocation -> {
+                  enteredWaitForAndReuseOutputs.countDown();
+                  completeWaitForAndReuseOutputs.await();
+                  return (SpawnResult) invocation.callRealMethod();
```
Comment (tjgq) [inline src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java:800]: As a sanity check, can we verify that this is only called for the second spawn?

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java (hunk 2 above)
Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/remote/RemoteSpawnCacheTest.java]: Done. Note that with the latch we can no longer assert that the thread would naturally wait (in the call to `awaitAllOutputReuse`), so I removed that assert.

**Review by tjgq — APPROVED** [review]

**Pull request closed without merging**